File: 01Singleton/Students/2017/KrygerDamian/src/main/java/pl/devdiary/wzorce/singleton/SerializationSafeSingleton.java
package pl.devdiary.wzorce.singleton;

import java.io.Serializable;

public class SerializationSafeSingleton implements Serializable {
    private static SerializationSafeSingleton instance;

    private SerializationSafeSingleton() {}

    public static SerializationSafeSingleton getInstance() {
        if(instance == null) {
            instance = new SerializationSafeSingleton();
        }

        return instance;
    }

    protected Object readResolve() {
        return getInstance();
    }
}

File: 01Singleton/Students/2017/ddziag/src/main/SerializationSafeSingleton.scala
package singleton

@SerialVersionUID(100L)
class SerializationSafeSingleton private(var state: Int) extends Serializable {
  def increment(): Unit = {
    state += 1
  }

  def readResolve(): Object = {
    SerializationSafeSingleton.instance
  }
}

object SerializationSafeSingleton {
  //Lazy values in Scala can hold null values. Access to lazy value is thread-safe.
  lazy val instance = new SerializationSafeSingleton(0)

}


